module/zfs/dbuf.c

```diff
@@ -2745,8 +2747,8 @@ dmu_buf_will_fill(dmu_buf_t *db_fake, dmu_tx_t *tx)
 		 */
 		mutex_enter(&db->db_mtx);
```

Comment:
It looks to me like we should be protecting the `if (db->db_state == DB_NOFILL)` check above as well.

Reply:
Good point. I updated the code to hold the `db_mtx` across the entire `if (db->db_state == DB_NOFILL)` code block.